R/estimate_incubation.R

```diff
@@ -0,0 +1,114 @@
+#' Function to extract the empirical incubation period distribution from line list data.
+#' Can take into account uncertain dates of exposure.
+#'
+#' @param x the linelist data (data.frame or linelist object) containing at least a column containing the exposure dates and one containing the onset dates. For exposure dates, each element can be a vector containing several possible exposure dates. Note that if the same exposure date appears twice in the list it si given twice as much weight.
+#' @param dates_exposure the name of the column containing the exposure dates (bare variable name or in quotes)
+#' @param date_of_onset the name of the column containing the onset dates (bare variable name or in quotes)
+#' @return a data frame containing a column with the different incubation periods and a column containing their relative frequency
+#' @author Flavio Finger, \email{flavio.finger@lshtm.ac.uk}
+#' @export
+#' @importFrom dplyr pull
+#' @importFrom rlang "!!" enquo
+#' @examples
+#' x <- linelist::clean_data(linelist::messy_data())
+#'
+#' mkexposures <- function(x) x - round(rgamma(sample.int(5, size = 1), shape = 12, rate = 3))
+#' exposures <- sapply(x$date_of_onset, mkexposures)
+#' x$dates_exposure <- exposures
+#'
+#' incubation_period_dist <- empirical_incubation_dist(x, dates_exposure, date_of_onset)
+
+empirical_incubation_dist  <- function(x, dates_exposure, date_of_onset) {
+  #error checking
+  if (!is.data.frame(x)) {
+    stop("x is not a data.frame")
+  }
+
+  if (ncol(x)==0L) {
+    stop("x has no columns")
+  }
+
+  dates_exposure <- rlang::enquo(dates_exposure)
+  date_of_onset <- rlang::enquo(date_of_onset)
+
+  y <- compute_incubation(dplyr::pull(x, !!dates_exposure), dplyr::pull(x, !!date_of_onset))
+
+  #check if incubation period is below 0
+  if (any(y$incubation_period < 0)){
+    warning("negative incubation periods in data!")
+  }
+
+  return(y)
+}
+
+#' Compute the empirical incubation dist.
+#' Can take into account uncertain dates of exposure.
+#'
+#' @param dates_exposure list containing the exposure dates. each element can be a vector of several possible exposure dates.
+#' @param date_onset list containing the exposure dates. each element can be a vector of several dates.
+#' @return a data frame containing a column with the different incubation periods and a column containing their relative frequency
+#' @author Flavio Finger, \email{flavio.finger@lshtm.ac.uk}
+#' @importFrom dplyr mutate select pull
+#' @importFrom rlang "!!"
+#' @importFrom purrr map
+#' @importFrom tidyr unnest
+#' @importFrom magrittr "%>%"
+compute_incubation <- function(dates_exposure, date_onset){
+  z <- data.frame(date_onset = date_onset)
+  z$dates_exposure <- dates_exposure
+  incubation_period <- NULL #entry to avoid note by R CMD check
```

Comment:
I'm not a big fan of this tactic. I would suggest to use this tactic:

```r
incubation_period <- quote(incubation_period)
weight <- quote(weight)

...

dplyr::select(!!incubation_period, !!weight)
```

Excruciating details here: https://cran.r-project.org/web/packages/dplyr/vignettes/programming.html